### ISSUE TITLE
Suggested changes to Parallel Coordinate chart

### ISSUE THREAD

**Issue by lakshaykc** [allegroai/clearml#259]
Here are a few suggestions to improve usability and the look & feel

1. Add a table below the chart to show the applied filters on the parallel coordinate figure. Say I have 10 experiments on the chart and I create a filter on one of the hyper-parameters which keeps only 4 experiments. There should be a table that shows those filtered experiments with all the hyper-parameters. The hyper parameter list might be long, so it would need to be a horizontally scrollable table. Or atleast it should display the hyper-params displayed on the figure.

    Why we need this?
If I'm comparing many experiments, it is hard to distinguish just based on the colors. Imagine finding an experiment from 10-20 colors. Also, it is useful to know other hyper-params of that task and not just the ones I'm comparing

2. Don't display the performance metric value on the chart. It gets too cluttered. When I have many experiments, they overlap each other and the information is useful. If the above feature is added, we wouldn't need to display it on the chart. We can just look it up in the table below.

3. Allow more than one performance metrics. I want to add both train_mse and valid_mse, but currently I can only add 1. Why is that?

4. When I make a selection on the chart, the filters get reset if I hover my mouse over the legend below the figure. I'm not sure if this is a bug or it intended. It is very easy to hover below the chart (either unintentionally or intentionally) and all the filters get reset which takes time to create again. Checkout the gif below for this.

![para_coord_1](https://user-images.githubusercontent.com/29871062/100488118-22775780-30c1-11eb-96b5-2bc9c8564f9e.gif)
 
5. Maybe we should be anti-aliasing the lines. It is visually more pleasing to the eye. While not exactly the same issue but checkout the difference in plots [here](https://github.com/plotly/plotly.js/issues/2756#issue-336233140)

**Comment by bmartinn** [comment]
Thanks @lakshaykc 

1. I think this was suggested, and hopefully we will get to implementing it, I can definitely see the value.
Meanwhile you can achieve some of the functionality with the experiment table and custom columns :slightly_smiling_face:

2.  "Don't display the performance metric" 
Isn't that important? what am I missing?

3. I'm assuming you mean to add additional columns to the right, correct? This is obviously doable, but it will make the metric column selection much more complicated. What I'm trying to gauge is the amount of value we get in return, what do you think?

4. This looks like a bug. Could you describe what are the exact steps to reproduce ? Is it just moving the mouse cursor outside the plot ?

5. Is this a plotly feature we can turn on? Is it a browser issue? What is browser you were testing with?

**Comment by lakshaykc** [comment]
> 1. I think this was suggested, and hopefully we will get to implementing it, I can definitely see the value.
Meanwhile you can achieve some of the functionality with the experiment table and custom columns 🙂

Ok, great. It is quite valuable when I have large number of experiments. When you say 'experiment table' and 'custom columns', do you mean the columns in the parallel coordinate chart? Which experiment table are you referring to in this case?

> 2. "Don't display the performance metric"
Isn't that important? what am I missing?

Well, when the number of experiments in the chart become too large, the numbers overlap each other and you can't read anything, so it is not useful. In the image above, we can still see it but there are only 10 experiments. But when I have more 20, 50, etc, it gets very cluttered. Maybe there would be threshold below which the numbers are displayed and above it, they aren't.

![Screen Shot 2020-12-01 at 9 21 48 PM](https://user-images.githubusercontent.com/29871062/100831765-43baa980-341b-11eb-8e33-0f0f8c98d48c.png)

I think this goes along with the point 1, where if we had the table of the selection which we could sort by any metric, we wouldn't need to display the metrics on the chart. What do you think?

> 3. I'm assuming you mean to add additional columns to the right, correct? This is obviously doable, but it will make the metric column selection much more complicated. What I'm trying to gauge is the amount of value we get in return, what do you think?

So this assessment might change, if we have the table implemented, right?! But let's say even without it, I would like to know additional metrics on the parallel coordinate chart which are not hyperparameters. Even if I'm judging the run by validation mse, I'd still like see the train mse. I have additional metrics that I log during the training which are helpful to see when comparing in the parallel coordinate chart. Ofcourse I can get those metrics by the looking at individual metrics, but again when comparing 10s of experiments, it becomes cumbersome. Does that make sense?

> 4. This looks like a bug. Could you describe what are the exact steps to reproduce ? Is it just moving the mouse cursor outside the plot ?

Yes, I just dragged filter bars on the chart and moved the mouse cursor down to where the experiment names are displayed.

> 5. Is this a plotly feature we can turn on? Is it a browser issue? What is browser you were testing with?

This is probably more of a plotly issue than trains. Let me research this out a little more and I'll get back to. I used to use this commercial software called modefrontier (worked there) which is not in the ML domain but has a very good visualization toolset. The software had an option to use `anti-aliasing` and with the option selected the visualization was dramatically better. Since the tool used plotly as its base, maybe there is an option. Also, it seem like Weights & Biases plots use anti-aliasing and they do look good.

**Comment by bmartinn** [comment]
>  When you say 'experiment table' and 'custom columns', do you mean the columns in the parallel coordinate chart? Which experiment table are you referring to in this case?

I mean the main experiment page in the UI (the one displaying all the experiments in the project).
If you press on the cogwheel icon, you can add any columns you like (basically the same information as you have in the comparison page: metrics, hyper-parameters etc.).
For example, this allows you to add a column for accuracy and a column for loss then sort by accuracy but sill get a full view of all the hyper parameters. Here is an example on the [demo server](https://demoapp.trains.allegro.ai/projects/9a3e959bd6154175876510c23cb7581f/experiments?columns=selected&columns=type&columns=name&columns=tags&columns=status&columns=users&columns=started&columns=last_update&columns=last_iteration&columns=p.Args.beta1&columns=m.81513effdf5790b79549208838404407.81513effdf5790b79549208838404407.min_value.loss.loss&columns=m.0dc7d8688b9eec1e71cda81967772c39.0dc7d8688b9eec1e71cda81967772c39.max_value.out_binary_accuracy.out_binary_accuracy&columns=p.Args.beta2&columns=p.Args.blur&columns=project.name&order=last_metrics.0dc7d8688b9eec1e71cda81967772c39.0dc7d8688b9eec1e71cda81967772c39.max_value)


2. So maybe display the value on when hovering ?

3. Make sense :)

4. Okay, I'll make sure we check it, I suspect this might also be a plotly bug... Anyhow we need to upgrade to the latest one.

5. Sounds good, keep me posted on what you find

**Comment by lakshaykc** [comment]
> 1.
Aah, got it.

> 2.
I think that would be better

**Comment by jkhenning** [comment]
Closing this as it was already released. Please reopen if required.
